File: BSvsZP-Common/CommonTester/BombTester.cs
using System;
using System.Collections.Generic;
using Microsoft.VisualStudio.TestTools.UnitTesting;

using Common;

namespace CommonTester
{
    [TestClass]
    public class BombTester
    {
        [TestMethod]
        public void Bomb_CheckConstructors()
        {
            Bomb b = new Bomb();
            Assert.AreEqual(0, b.CreatorId);
            Assert.IsNotNull(b.Excuses);
            Assert.IsNotNull(b.Twine);
            Assert.AreEqual(0, b.Excuses.Count);
            Assert.AreEqual(0, b.Twine.Count);
            Assert.IsNull(b.BuiltOnTick);

            Tick t1 = new Tick();
            List<Excuse> eList = new List<Excuse> { CreateExcuse(10), CreateExcuse(11), CreateExcuse(12) };
            List<WhiningTwine> wtList = new List<WhiningTwine> { CreateTwine(20), CreateTwine(21), CreateTwine(22) };
            b = new Bomb(1, eList, wtList, t1);
            Assert.AreEqual(1, b.CreatorId);
            Assert.AreSame(eList, b.Excuses);
            Assert.AreSame(wtList, b.Twine);
            Assert.AreSame(t1, b.BuiltOnTick);
        }

        [TestMethod]
        public void Bomb_CheckProperties()
        {
            Tick t1 = new Tick();
            List<Excuse> eList = new List<Excuse> { CreateExcuse(10), CreateExcuse(11), CreateExcuse(12) };
            List<WhiningTwine> wtList = new List<WhiningTwine> { CreateTwine(20), CreateTwine(21), CreateTwine(22) };
            Bomb b = new Bomb(1, eList, wtList, t1);
            Assert.AreEqual(1, b.CreatorId);
            Assert.AreSame(eList, b.Excuses);
            Assert.AreSame(wtList, b.Twine);
            Assert.AreSame(t1, b.BuiltOnTick);

            b.CreatorId = 135;
            Assert.AreEqual(135, b.CreatorId);
            b.CreatorId = 0;
            Assert.AreEqual(0, b.CreatorId);
            b.CreatorId = Int16.MaxValue;
            Assert.AreEqual(Int16.MaxValue, b.CreatorId);

            b.Excuses = new List<Excuse>();
            Assert.IsNotNull(b.Excuses);
            Assert.AreNotSame(eList, b.Excuses);
            b.Excuses = eList;
            Assert.AreSame(eList, b.Excuses);

            b.Twine = new List<WhiningTwine>();
            Assert.IsNotNull(b.Twine);
            Assert.AreNotSame(wtList, b.Twine);
            b.Twine = wtList;
            Assert.AreSame(wtList, b.Twine);
        }

        [TestMethod]
        public void Bomb_CheckEncodeAndDecode()
        {
            Tick t1 = new Tick();
            List<Excuse> eList = new List<Excuse> { CreateExcuse(10), CreateExcuse(11), CreateExcuse(12) };
            List<WhiningTwine> wtList = new List<WhiningTwine> { CreateTwine(20), CreateTwine(21), CreateTwine(22) };
            Bomb b1 = new Bomb(1, eList, wtList, t1);
            Assert.AreEqual(1, b1.CreatorId);
            Assert.AreSame(eList, b1.Excuses);
            Assert.AreSame(wtList, b1.Twine);
            Assert.AreSame(t1, b1.BuiltOnTick);

            ByteList bytes = new ByteList();
            b1.Encode(bytes);
            Bomb b2 = Bomb.Create(bytes);
            Assert.AreEqual(b1.CreatorId, b2.CreatorId);
            Assert.AreEqual(b1.Excuses.Count, b2.Excuses.Count);
            Assert.AreEqual(b1.Twine.Count, b2.Twine.Count);

            Assert.AreEqual(b1.BuiltOnTick.LogicalClock, b2.BuiltOnTick.LogicalClock);
            Assert.AreEqual(b1.BuiltOnTick.HashCode, b2.BuiltOnTick.HashCode);

            bytes.Clear();
            b1.Encode(bytes);
            bytes.GetByte();            // Read one byte, which will throw the length off
            try
            {
                b2 = Bomb.Create(bytes);
                Assert.Fail("Expected an exception to be thrown");
            }
            catch (ApplicationException)
            {
            }

            bytes.Clear();
            b1.Encode(bytes);
            bytes.Add((byte)100);       // Add a byte
            bytes.GetByte();            // Read one byte, which will make the ID wrong
            try
            {
                b2 = Bomb.Create(bytes);
                Assert.Fail("Expected an exception to be thrown");
            }
            catch (ApplicationException)
            {
            }
        }

        private Excuse CreateExcuse(Int16 id)
        {
            return new Excuse(id, new List<Tick> { new Tick(), new Tick() }, new Tick());
        }

        private WhiningTwine CreateTwine(Int16 id)
        {
            return new WhiningTwine(id, new List<Tick> { new Tick(), new Tick() }, new Tick());
        }

    }
}

File: BSvsZP-Common -Java/CommonTester/BombTester.java
package CommonTester;

import static org.junit.Assert.*;

import java.net.UnknownHostException;
import java.util.ArrayList;
import java.util.Arrays;

import org.junit.Test;
import org.omg.CORBA.portable.ApplicationException;

import Common.*;

public class BombTester {

	@Test
	public void Bomb_CheckConstructors()
    {
        Bomb b = new Bomb();
        assertEquals(0, b.CreatorId);
        assertNotNull(b.Excuses);
        assertNotNull(b.Twine);
        assertEquals(0, b.Excuses.size());
        assertEquals(0, b.Twine.size());
        assertNull(b.BuiltOnTick);

        Tick t1 = new Tick();
        ArrayList<Excuse> eList = new ArrayList<Excuse>();
        eList.add(CreateExcuse((short) 10));
        eList.add(CreateExcuse((short) 11));
        eList.add(CreateExcuse((short) 12));
        
        ArrayList<WhiningTwine> wtList = new ArrayList<WhiningTwine>();
        wtList.add(CreateTwine((short) 20));
        wtList.add(CreateTwine((short) 21));
        wtList.add(CreateTwine((short) 22));
        
        b = new Bomb((short) 1, eList, wtList, t1);
        assertEquals(1, b.CreatorId);
        assertSame(eList, b.Excuses);
        assertSame(wtList, b.Twine);
        assertSame(t1, b.BuiltOnTick);
    }
	
	@Test
	public void Bomb_CheckProperties()
    {
        Tick t1 = new Tick();
        
        ArrayList<Excuse> eList = new ArrayList<Excuse>
		(Arrays.asList(CreateExcuse((short) 10),
					   CreateExcuse((short) 11), 
					   CreateExcuse((short) 12)));
        
        ArrayList<WhiningTwine> wtList = new ArrayList<WhiningTwine>
		(Arrays.asList(CreateTwine((short) 20),
		  			   CreateTwine((short) 21),
					   CreateTwine((short) 22)));
        
        
        Bomb b = new Bomb((short)1, eList, wtList, t1);
        assertEquals(1, b.CreatorId);
        assertSame(eList, b.Excuses);
        assertSame(wtList, b.Twine);
        assertSame(t1, b.BuiltOnTick);

        b.CreatorId = 135;
        assertEquals(135, b.CreatorId);
        b.CreatorId = 0;
        assertEquals(0, b.CreatorId);
        b.CreatorId = Short.MAX_VALUE;
        assertEquals(Short.MAX_VALUE, b.CreatorId);

        b.Excuses = new ArrayList<Excuse>();
        assertNotNull(b.Excuses);
        assertNotSame(eList, b.Excuses);
        b.Excuses = eList;
        assertSame(eList, b.Excuses);

        b.Twine = new ArrayList<WhiningTwine>();
        assertNotNull(b.Twine);
        assertNotSame(wtList, b.Twine);
        b.Twine = wtList;
        assertSame(wtList, b.Twine);
    }

	@Test
	 public void Bomb_CheckEncodeAndDecode() throws Exception
    {
        Tick t1 = new Tick();
        ArrayList<Excuse> eList = new ArrayList<Excuse>
        		(Arrays.asList(CreateExcuse((short) 10),
        					   CreateExcuse((short) 11), 
        					   CreateExcuse((short) 12)));
        
        ArrayList<WhiningTwine> wtList = new ArrayList<WhiningTwine>
        		(Arrays.asList(CreateTwine((short) 20),
        		  			   CreateTwine((short) 21),
        					   CreateTwine((short) 22)));
        
        Bomb b1 = new Bomb((short) 1, eList, wtList, t1);
        
        assertEquals(1, b1.CreatorId);
        assertSame(eList, b1.Excuses);
        assertSame(wtList, b1.Twine);
        assertSame(t1, b1.BuiltOnTick);

        ByteList bytes = new ByteList();
        b1.Encode(bytes);
        Bomb b2 = Bomb.Create(bytes);
        assertEquals(b1.CreatorId, b2.CreatorId);
        assertEquals(b1.Excuses.size(), b2.Excuses.size());
        assertEquals(b1.Twine.size(), b2.Twine.size());
        assertEquals(b1.BuiltOnTick.getLogicalClock(), b2.BuiltOnTick.getLogicalClock());
        assertEquals(b1.BuiltOnTick.getHashCode(), b2.BuiltOnTick.getHashCode());

        bytes.Clear();
        b1.Encode(bytes);
        bytes.GetByte();            // Read one byte, which will throw the length off
        try
        {
        	b2 = Bomb.Create(bytes);
        	fail("Expected an exception to be thrown");
        } catch (Exception e){}

        bytes.Clear();
        b1.Encode(bytes);
        bytes.Add((byte)100);       // Add a byte
        bytes.GetByte();            // Read one byte, which will make the ID wrong
        
        try
        {
        	b2 = Bomb.Create(bytes);
        	fail("Expected an exception to be thrown");
        } catch (Exception e) {}
        
    }
	
	private Excuse CreateExcuse(short id)
    {
        return new Excuse(id, new ArrayList<Tick>(Arrays.asList(new Tick(), new Tick())), new Tick());
    }

    private WhiningTwine CreateTwine(short id)
    {
    	return new WhiningTwine(id, new ArrayList<Tick>(Arrays.asList(new Tick(), new Tick())), new Tick());
    }



}
